src/libraries/System.Runtime.Serialization.BinaryFormat/src/System/Runtime/Serialization/BinaryFormat/RectangularOrCustomOffsetArrayRecord.cs

```diff
@@ -0,0 +1,239 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Reflection.Metadata;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.Serialization.BinaryFormat.Utils;
+
+namespace System.Runtime.Serialization.BinaryFormat;
+
+internal sealed class RectangularOrCustomOffsetArrayRecord : ArrayRecord
+{
+    private readonly int[] _lengths;
+    private readonly ICollection<object> _values;
+    private TypeName? _elementTypeName;
+
+    private RectangularOrCustomOffsetArrayRecord(Type elementType, ArrayInfo arrayInfo,
+        MemberTypeInfo memberTypeInfo, int[] lengths, int[] offsets, bool canPreAllocate) : base(arrayInfo)
+    {
+        ElementType = elementType;
+        MemberTypeInfo = memberTypeInfo;
+        _lengths = lengths;
+        Offsets = offsets;
+
+        // A List<T> can hold as many objects as an array, so for multi-dimensional arrays
+        // with more elements than Array.MaxLength we use LinkedList.
+        // Testing that many elements takes a LOT of time, so to ensure that both code paths are tested,
+        // we always use LinkedList code path for non-Release builds.
+#if RELEASE
+        _values = arrayInfo.TotalElementsCount <= ArrayInfo.MaxArrayLength
+            ? new List<object>(canPreAllocate ? arrayInfo.GetSZArrayLength() : Math.Min(4, arrayInfo.GetSZArrayLength()))
+            : new LinkedList<object>();
+#else
+        _values = new LinkedList<object>();
+#endif
```

Comment:
I don't recall, off the top of my head, ever seeing `#if RELEASE`.  Usually it's the `DEBUG` preprocessor symbol that gets tested.

```suggestion
        // we always use LinkedList code path for Debug builds.
#if DEBUG
        _values = new LinkedList<object>();
#else
        _values = arrayInfo.TotalElementsCount <= ArrayInfo.MaxArrayLength
            ? new List<object>(canPreAllocate ? arrayInfo.GetSZArrayLength() : Math.Min(4, arrayInfo.GetSZArrayLength()))
            : new LinkedList<object>();
#endif
```